src/rgw/rgw_notify.cc

```diff
@@ -2,20 +2,596 @@
+      const auto push_endpoint = RGWPubSubEndpoint::create(record_with_endpoint.push_endpoint, record_with_endpoint.arn_topic,
+          RGWHTTPArgs(record_with_endpoint.push_endpoint_args), 
+          cct);
+      ldout(cct, 20) << "INFO: push endpoint created: " << record_with_endpoint.push_endpoint <<
+        " for entry: " << entry.marker << dendl;
+      const auto ret = push_endpoint->send_to_completion_async(cct, record_with_endpoint.record, optional_yield(io_context, yield));
+      if (ret < 0) {
+        ldout(cct, 5) << "WARNING: push entry: " << entry.marker << " to endpoint: " << record_with_endpoint.push_endpoint 
+          << " failed. error: " << ret << " (will retry)" << dendl;
+        return false;
+      } else {
+        ldout(cct, 20) << "INFO: push entry: " << entry.marker << " to endpoint: " << record_with_endpoint.push_endpoint 
+          << " ok" <<  dendl;
+        if (perfcounter) perfcounter->inc(l_rgw_pubsub_push_ok);
+        return true;
+      }
+    } catch (const RGWPubSubEndpoint::configuration_error& e) {
+      ldout(cct, 5) << "WARNING: failed to create push endpoint: " 
+          << record_with_endpoint.push_endpoint << " for entry: " << entry.marker << ". error: " << e.what() << " (will retry) " << dendl;
+      return false;
+    }
+  }
+
+  bool queue_owned(const std::string& queue_name, spawn::yield_context yield) {
+    librados::ObjectWriteOperation op;
+    op.assert_exists();
+    rados::cls::lock::lock(&op, queue_name+"_lock", 
+        ClsLockType::EXCLUSIVE,
+        lock_cookie, 
+        "" /*no tag*/,
+        "" /*no description*/,
+        failover_time,
+        LOCK_FLAG_MUST_RENEW);
+
+    const auto ret = rgw_rados_operate(rados_ioctx, queue_name, &op, optional_yield(io_context, yield));
+    if (ret == -EBUSY) {
+      ldout(cct, 5) << "WARNING: queue: " << queue_name << " ownership moved to another daemon. processing will stop" << dendl;
+      return false;
+    }
+    if (ret == -ENOENT) {
+      ldout(cct, 10) << "INFO: check ownership of queue: " << queue_name << " - already deleted" << dendl;
+      return false;
+    }
+    if (ret < 0) {
+      ldout(cct, 1) << "ERROR: check ownership of queue: " << queue_name << ". error: " << ret << dendl;
+      return true; 
+    }
+    return true;
+  }
+
+  // clean stale reservation from queue
+  void cleanup_queue(const std::string& queue_name, spawn::yield_context yield) {
+    while (queue_owned(queue_name, yield)) {
+      ldout(cct, 20) << "INFO: trying to perform stale reservation cleanup for queue: " << queue_name << dendl;
+      const auto now = ceph::coarse_real_time::clock::now();
+      const auto stale_time = now - std::chrono::seconds(stale_reservations_period_s);
+      librados::ObjectWriteOperation op;
+      op.assert_exists();
+      cls_2pc_queue_expire_reservations(op, stale_time);
+      auto ret = rgw_rados_operate(rados_ioctx, queue_name, &op, optional_yield(io_context, yield));
```

Comment:
two reasons:
* I tried to keep the 2pc_queue generic (though it is used only here...), adding a build in lock makes sense here, but not sure it makes sense in the general case
* lock is also used when creating the queue (see line 388 - 417) and its coroutine. so, the lock cannot be embedded completely inside the queue

one option (that may address both issues), is to have the lock created outside thew queue and provide its name as an optional parameter to the queue creation, if provided, it will be used inside the cls operations. @cbodley please let me know what you think?

Reply:
as discussed with @cbodley the 2pc_queue API will not change.
instead, asserting on the lock will be combined (in the same operation) with the 3 places in the code where actions (read or writes) on the queue are taken: 
* read the entry list
* delete acked entries
* reservation cleanup